### PR TITLE
Replace occurrences of on_before_accelerator_backend_setup_called with setup hook 

### PR DESCRIPTION
**Pull request by krishnakalyan3**
## What does this PR do?

Part-1/2  #11559

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by krishnakalyan3** [comment]
Fixes #11559

File: docs/source/extensions/callbacks.rst [hunk 1]
```diff
@@ -216,12 +216,6 @@ on_configure_sharded_model
 .. automethod:: pytorch_lightning.callbacks.Callback.on_configure_sharded_model
     :noindex:
 
-on_before_accelerator_backend_setup
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-.. automethod:: pytorch_lightning.callbacks.Callback.on_before_accelerator_backend_setup
-    :noindex:
-
```
Comment (ananthsub) [inline docs/source/extensions/callbacks.rst:224]: Don't remove this from the docs. Instead add to the docstring that this hook is deprecated, and it'll render on the docs site

File: pytorch_lightning/trainer/callback_hook.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ class TrainerCallbackHookMixin(ABC):
     callbacks: List[Callback] = []
     lightning_module: "pl.LightningModule"
 
-    def on_before_accelerator_backend_setup(self) -> None:
+    def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/callback_hook.py:39]: this should not be changed. this was the old convention for how callback hooks were called

**Review by ananthsub — CHANGES_REQUESTED** [review]
Thanks for working on this @krishnakalyan3 

I'd make this into 2 PRs:
1. Transition all Lightning callbacks to use `setup` instead of `on_before_accelerator_backend_setup`. These callsites are shared in the initial issue
2. Mark the callback hook as deprecated (adding to the config validator, change the docstring of the hook to reflect it's deprecated, etc)

**Comment by krishnakalyan3** [comment]
@ananthsub These are the callbacks referenced by `on_before_accelerator_backend_setup`. Their functions signature already looks correct according to me. I felt that those changes were minimal. Can you please let me know if I missed out on something? If yes I would be happy to create another PR for it.

```
./pytorch_lightning/callbacks/pruning.py:    def on_before_accelerator_backend_setup(self, trainer: "pl.Trainer", pl_module: LightningModule) -> None:
./pytorch_lightning/callbacks/finetuning.py:    def on_before_accelerator_backend_setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
./pytorch_lightning/callbacks/stochastic_weight_avg.py:    def on_before_accelerator_backend_setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
./pytorch_lightning/callbacks/lambda_function.py:        on_before_accelerator_backend_setup: Optional[Callable] = None,
./pytorch_lightning/callbacks/base.py:    def on_before_accelerator_backend_setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:```

File: tests/callbacks/test_finetuning_callback.py [hunk 3]
```diff
@@ -268,7 +268,7 @@ def test_on_before_accelerator_backend_setup(tmpdir):
     configure_optimizers function call."""
 
     class TestCallback(Callback):
-        def on_before_accelerator_backend_setup(self, trainer, pl_module):
+        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (ananthsub) [inline tests/callbacks/test_finetuning_callback.py:271]:
```suggestion
        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
```

**Review by ananthsub — COMMENTED** [review]
This needs to be updated too: https://github.com/PyTorchLightning/pytorch-lightning/blob/76c23f34b889fdce3351d1b1c2f657d52ff95bb2/tests/callbacks/test_stochastic_weight_avg.py#L236-L241

**Comment by krishnakalyan3** [comment]
Since this is a minor change do I need to update the change logs?.

**Comment by daniellepintz** [comment]
A few comments:
1. It looks like you are missing references in two tests: `test_swa_deepcopy` and `test_deepspeed_assert_config_zero_offload_disabled`
2.  no need to update the changelog for this, but in the next PR yes
3. please update the title and description of this PR. this PR no longer fixes the issue directly, that will be the next PR

**Comment by krishnakalyan3** [comment]
@daniellepintz Let me know if the new changes look good?.

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 4]
```diff
@@ -233,8 +234,8 @@ def __init__(self, *args, **kwargs):
             super().__init__(*args, **kwargs)
             self.on_before_accelerator_backend_setup_called = False
 
-        def on_before_accelerator_backend_setup(self, trainer: "Trainer", pl_module: "LightningModule"):
-            super().on_before_accelerator_backend_setup(trainer, pl_module)
+        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
+            super().__init__(*args, **kwargs)
```
Comment (daniellepintz) [inline tests/callbacks/test_stochastic_weight_avg.py:238]:
```suggestion
            super().setup(trainer, pl_module, stage)
```

**Comment by daniellepintz** [comment]
@krishnakalyan3 Looks good! Two things:
1. I see a few references of `on_before_accelerator_backend_setup_called` across the tests, this should be updated to `setup_called`
2. Please update the title to "Replace occurrences of `on_before_accelerator_backend_setup_called` with `setup` hook"

**Comment by krishnakalyan3** [comment]
@daniellepintz let me know if this looks better.

**Review by daniellepintz — APPROVED** [review]
LTGM! Please fix the lint errors. You can check lint locally by running `pre-commit` (https://pre-commit.com/)

File: tests/callbacks/test_finetuning_callback.py [hunk 5]
```diff
@@ -268,16 +269,16 @@ def test_on_before_accelerator_backend_setup(tmpdir):
     configure_optimizers function call."""
 
     class TestCallback(Callback):
-        def on_before_accelerator_backend_setup(self, trainer, pl_module):
-            pl_module.on_before_accelerator_backend_setup_called = True
+        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
```
Comment (awaelchli) [inline tests/callbacks/test_finetuning_callback.py:272]: that looks fine to me, just fyi we usually don't add typing to the tests as it adds very little value

**Review by awaelchli — APPROVED** [review]

**Comment by krishnakalyan3** [comment]
@awaelchli let me know if the new changes look okay?.

**Comment by krishnakalyan3** [comment]
@daniellepintz 

Looks like the pre-commit messages are picking up this 
```
tests/strategies/test_deepspeed_strategy.py:409: [F821] undefined name 'pl'
tests/strategies/test_deepspeed_strategy.py:409: [F821] undefined name 'pl'
tests/callbacks/test_finetuning_callback.py:271: [F821] undefined name 'pl'
tests/callbacks/test_finetuning_callback.py:271: [F821] undefined name 'pl'
tests/callbacks/test_stochastic_weight_avg.py:236: [F821] undefined name 'pl'
tests/callbacks/test_stochastic_weight_avg.py:236: [F821] undefined name 'pl'
```

**Comment by awaelchli** [comment]
add `import pytorch_lightning as pl `
:)

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 6]
```diff
@@ -20,7 +20,8 @@
 from torch.optim.swa_utils import SWALR
 from torch.utils.data import DataLoader
 
-from pytorch_lightning import LightningModule, Trainer
+import pytorch_lightning as pl
```
Comment (rohitgr7) [inline tests/callbacks/test_stochastic_weight_avg.py:23]:
```suggestion
```

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 7]
```diff
@@ -231,20 +232,20 @@ def test_swa_deepcopy(tmpdir):
     class TestSWA(StochasticWeightAveraging):
         def __init__(self, *args, **kwargs):
             super().__init__(*args, **kwargs)
-            self.on_before_accelerator_backend_setup_called = False
+            self.setup_called = False
 
-        def on_before_accelerator_backend_setup(self, trainer: "Trainer", pl_module: "LightningModule"):
-            super().on_before_accelerator_backend_setup(trainer, pl_module)
+        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (rohitgr7) [inline tests/callbacks/test_stochastic_weight_avg.py:237]:
```suggestion
        def setup(self, trainer, pl_module, stage=None) -> None:
```

File: tests/strategies/test_deepspeed_strategy.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
 from torch.utils.data import DataLoader
 from torchmetrics import Accuracy
 
+import pytorch_lightning as pl
```
Comment (rohitgr7) [inline tests/strategies/test_deepspeed_strategy.py:16]:
```suggestion
```

File: tests/strategies/test_deepspeed_strategy.py [hunk 9]
```diff
@@ -406,7 +407,7 @@ def test_deepspeed_assert_config_zero_offload_disabled(tmpdir, deepspeed_zero_co
     deepspeed_zero_config["zero_optimization"]["offload_optimizer"] = False
 
     class TestCallback(Callback):
-        def on_before_accelerator_backend_setup(self, trainer, pl_module) -> None:
+        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (rohitgr7) [inline tests/strategies/test_deepspeed_strategy.py:410]:
```suggestion
        def setup(self, trainer, pl_module, stage=None) -> None:
```

File: tests/callbacks/test_finetuning_callback.py [hunk 10]
```diff
@@ -19,6 +19,7 @@
 from torch.optim import Optimizer, SGD
 from torch.utils.data import DataLoader
 
+import pytorch_lightning as pl
```
Comment (rohitgr7) [inline tests/callbacks/test_finetuning_callback.py:22]:
```suggestion
```

File: tests/callbacks/test_finetuning_callback.py [hunk 11]
```diff
@@ -268,16 +269,16 @@ def test_on_before_accelerator_backend_setup(tmpdir):
     configure_optimizers function call."""
 
     class TestCallback(Callback):
-        def on_before_accelerator_backend_setup(self, trainer, pl_module):
-            pl_module.on_before_accelerator_backend_setup_called = True
+        def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (rohitgr7) [inline tests/callbacks/test_finetuning_callback.py:272]:
```suggestion
        def setup(self, trainer, pl_module, stage=None) -> None:
```

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 12]
```diff
@@ -231,20 +232,20 @@ def test_swa_deepcopy(tmpdir):
...
+            super().setup(trainer, pl_module)
```
Comment (rohitgr7) [inline tests/callbacks/test_stochastic_weight_avg.py:238]:
```suggestion
            super().setup(trainer, pl_module, stage)
```

File: tests/callbacks/test_stochastic_weight_avg.py (hunk 7 above)
Comment (rohitgr7) [reply tests/callbacks/test_stochastic_weight_avg.py:237]: setup needs a stage parameter right?

File: tests/callbacks/test_stochastic_weight_avg.py (hunk 6 above)
Comment (rohitgr7) [reply tests/callbacks/test_stochastic_weight_avg.py:23]: typing isn't required.

**Comment by krishnakalyan3** [comment]
@rohitgr7 since `stage` is not referenced anywhere don't we have to initialize it somewhere as a global or a local variable?.

**Comment by rohitgr7** [comment]
> 

@krishnakalyan3 `stage` will be set up trainer and passed to the callback's `setup` internally.
for tests, typing isn't required.

**Review by rohitgr7 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
